Add size statistics to `ParquetMetaData` introduced in PARQUET-2261

etseidl opened the pull request:
# Which issue does this PR close?
<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #5022

# Rationale for this change
 
Implements new page and column chunk size statistics introduced in [PARQUET-2261](https://github.com/apache/parquet-format/pull/197)
<!--
Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

# What changes are included in this PR?

Adds the necessary structures from the updated `parquet.thrift`, and adds the code necessary to populate them.

# Are there any user-facing changes?

No
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please add the `breaking change` label.
-->

etseidl commented:
This is my first foray into Rust programming, so I'm not sure everything is done as idiomatically as possible. Submitting this now to get early feedback on my approach. I'm also wondering how much testing to add, and whether this should have a configuration parameter to turn generating the statistics off.

alamb commented:
I triggered the CI checks and will try and get to review this over the next few days if no one beats me to it

Samrose-Ahmed commented:
Hey @etseidl are you still working on this, otherwise I had some code in a fork, I can try to move this along.

alamb commented:
It seems I never looked at this PR but I just skimmed it and it looks reasonable to me -- I think the biggest thing it is currently lacking is tests

etseidl commented:
Yeah, sorry, I got sidetracked by other work...and unit tests are the bane of my existence 😅. Let me see if I can get some added in the next few days.

etseidl commented:
@alamb I've added two tests so far. I still need to test the repetition level histogram, but I think I've pushed `test_roundtrip()` in `writer.rs` as far as I can. Can you look at what I've done so far and see if it's acceptable?

I'm also not entirely sure how to test the new statistics in the page indexes. For now my story is if the `ColumnMetaData::SizeStatistics` are correct, then the per-page stats in the page indexes must also be correct.

etseidl commented:
I think this is ready, but just noticed #5854, which will impact this PR.

alamb commented:
> I think this is ready, but just noticed #5854, which will impact this PR.

I don't think we have specific plans to actually implement a custom thrift decoder at this time. That ticket came out work  @XiangpengHao did while writing   https://www.influxdata.com/blog/how-good-parquet-wide-tables/ -- to provide some background if someone really cares about  parquet metadata performance

alamb commented on parquet/src/file/metadata.rs:
```diff
@@ -665,6 +668,21 @@ impl ColumnChunkMetaData {
         Some(offset..(offset + length))
     }
 
+    /// Returns the number of bytes of variable length data.
```
```suggestion
    /// Returns the number of bytes of variable length data after decoding
```

alamb commented on parquet/src/file/metadata.rs:
```diff
@@ -665,6 +668,21 @@ impl ColumnChunkMetaData {
+        self.unencoded_byte_array_data_bytes
+    }
+
+    /// Returns the repetition level histogram.
```
Perhaps we can inline some information on what the histogram contains.

Same comment applies to `definition_level_histograms`

```suggestion
    /// Returns the repetition level histogram.
    ///
    /// The returned value `vec[i]` is how many rows are at repetition level `i`
```

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -1828,4 +1836,93 @@ mod tests {
+        let unenc_size: i64 = data[0].iter().map(|x| x.len() as i64).sum();
+        let file: File = tempfile::tempfile().unwrap();
+
+        let file_metadata = test_roundtrip::<File, File, ByteArrayType, _>(
```
Could we please add some test that verifies that the metadata read from the file correctly interprets the new statistics? 

It seems to me like `test_roundtrip` only verifies a few fields of the metadata after it is read back in

https://github.com/apache/arrow-rs/blob/73fa0994bbb2ba1795fdb3b050a66ca59864f403/parquet/src/file/writer.rs#L1461-L1482

etseidl replied:
Yes, file metadata returned from `test_roundtrip` is created by the writer, not the reader. So I'm currently just testing that the thrift objects are properly formed before writing. The issue I have is that the reader doesn't currently save the new structures, and I don't want to add them on this pass since they won't be used at all. I can instead replicate some of the code in `index_reader.rs` to create thrift objects from the file data and test that those match the objects created by the writer.

alamb replied:
🤔  it seems like it would be good to have the ability to read the structures as well (otherwise pure rust implementations can't take advantage of these fields )

etseidl replied:
I agree (and in fact I know people using DataFusion that want these stats...but are ok reading them directly from the files as a preprocessing step). I think adding the histograms to `Index` is easy enough, but currently the `OffsetIndex` is returned as a vector of `PageLocation`s. Are you ok with me adding an object that mirrors the thrift `OffsetIndex`?

Samrose-Ahmed replied:
Yes I think we should add add our own `SizeStatistics` struct `src/file/size_statistics.rs` that mirrors the Thrift TSizeStatistics.

alamb replied:
The thing we need to be careful about is not introducing any breaking API changes or else we can't merge this PR until August 2024 per https://github.com/apache/arrow-rs?tab=readme-ov-file#release-versioning-and-schedule

etseidl replied:
Hmm. Adding histograms to `Index` won't be a breaking change as they'll be optional. I could add another method to `index_reader.rs` to return a vector of unencoded byte array sizes. It's less efficient since we'd have to parse the offset index twice.

`ColumnChunkMetaData` contains the chunk level histograms and byte array sizes already, just not as a separate `SizeStatistics` object.

alamb reviewed (commented):
Thanks @etseidl  -- in general I think this looks good to me. My only real concern is that it will add additional overhead to parquet writing (especially as there is a per-row calculation for the histograms)

I'll run some benchmarks now

If we see any slowdown in the encoding, I think we should make writing this metadata optional 

Some method on https://docs.rs/parquet/latest/parquet/file/properties/struct.WriterProperties.html

It seems like https://docs.rs/parquet/latest/parquet/file/properties/struct.WriterProperties.html#method.statistics_enabled is the natural choice, but I am not sure how to extend that enum to capture this

alamb commented:
🤔  now that I look at the benchmarks there don't seem to be any that test the encoding of the entire file 🤔

alamb commented:
I basically ran 

```shell
cargo bench --bench arrow_writer --all-features
```

And here are the results. My conclusion is that there is no measurable difference 


```
group                                                                     master                                 size_stats
-----                                                                     ------                                 ----------
write_batch nested/4096 values primitive list                             1.00      2.2±0.07ms    72.7 MB/sec    1.02      2.3±0.04ms    71.4 MB/sec
write_batch nested/4096 values primitive list non-null                    1.00      2.7±0.05ms    70.6 MB/sec    1.01      2.7±0.06ms    69.6 MB/sec
write_batch primitive/4096 values bool                                    1.02   439.0±10.72µs     2.4 MB/sec    1.00   429.3±33.40µs     2.5 MB/sec
write_batch primitive/4096 values bool non-null                           1.04   362.9±19.33µs  1614.5 KB/sec    1.00   349.5±24.45µs  1676.6 KB/sec
write_batch primitive/4096 values primitive                               1.02      2.2±0.05ms    81.2 MB/sec    1.00      2.1±0.04ms    82.5 MB/sec
write_batch primitive/4096 values primitive non-null                      1.02      2.0±0.18ms    85.7 MB/sec    1.00  1965.6±98.22µs    87.8 MB/sec
write_batch primitive/4096 values primitive non-null with bloom filter    1.00     30.0±0.60ms     5.7 MB/sec    1.00     30.0±0.40ms     5.7 MB/sec
write_batch primitive/4096 values primitive with bloom filter             1.00     29.8±0.39ms     5.9 MB/sec    1.00     29.8±0.51ms     5.9 MB/sec
write_batch primitive/4096 values string                                  1.00  1159.0±50.52µs   108.9 MB/sec    1.00  1153.5±56.50µs   109.4 MB/sec
write_batch primitive/4096 values string dictionary                       1.00   738.6±18.18µs    64.8 MB/sec    1.02   754.8±23.34µs    63.4 MB/sec
write_batch primitive/4096 values string dictionary with bloom filter     1.00      4.3±0.04ms    11.1 MB/sec    1.00      4.3±0.01ms    11.1 MB/sec
write_batch primitive/4096 values string non-null                         1.00  1301.4±48.63µs    60.3 MB/sec    1.02  1321.0±78.28µs    59.4 MB/sec
write_batch primitive/4096 values string non-null with bloom filter       1.00      8.7±0.06ms     9.1 MB/sec    1.00      8.7±0.07ms     9.1 MB/sec
write_batch primitive/4096 values string with bloom filter                1.00      8.6±0.07ms     9.3 MB/sec    1.00      8.6±0.04ms     9.3 MB/sec
```

alamb commented:
Thank you @etseidl  -- this is a fairly impressive first foray into Rust programming. Sorry it took so long to review

etseidl commented:
Thanks for the great feedback @alamb. I'll get started on trying to get the stats in the read path and improve the testing.

etseidl commented:
Histograms are added to the column index. Tomorrow I'll work on the offset index.

etseidl commented:
@alamb I have the offset index done now well enough to test the metadata (might still need doc updates). I'm not entirely happy with it since the offset index is parsed twice now. I'm fine with either proceeding with this and then submitting a follow-on PR to add a proper offset index struct to hold the page locations and unencoded array sizes, or just doing both in this PR and waiting until 53.0.0.

etseidl commented on parquet/src/column/writer/encoder.rs:
```diff
@@ -150,6 +152,11 @@ impl<T: DataType> ColumnValueEncoderImpl<T> {
             }
+
+            if let Some(var_bytes) = T::T::variable_length_bytes(slice) {
+                self.variable_length_bytes =
+                    Some(var_bytes + self.variable_length_bytes.unwrap_or(0));
```
Is this better as:
```suggestion
                *self.variable_length_bytes.get_or_insert(0) += var_bytes;
```

alamb commented:
Thanks @etseidl  -- I will try and find time this week to review this PR in more detail

alamb reviewed (commented):
Thank you @etseidl -- this is a great contribution. 

I left some code structural suggestions that I think would help make the code easier to understand. 

I also think for testing we should have at least one end to end test (perhaps in arrow_writer) that writes out a parquet file, and then reads the metadata back in and verifies the contents

If you are willing, I would like to suggest we break this code down into a few PRs:
1. update parquet-format / add the newly generated code
2. Add support for `variable_length_bytes` (and change `ParquetColumnIndex`)`
3. Add support for rep/definition level histograms, update `ParquetOffsetIndex`

etseidl commented:
Thanks for the detailed and actionable review @alamb! I really appreciate you taking the time to review so thoroughly.

> If you are willing, I would like to suggest we break this code down into a few PRs:

Certainly. I can submit the format changes first, but I'll keep this open for the time being while I address your comments. Should I open new issues for each PR, or is it sufficient to reference #5022 for all three?

etseidl commented on parquet/src/format.rs:
```diff
@@ -634,6 +634,143 @@ impl From<&BoundaryOrder> for i32 {
   }
 }
 
+//
```
I manually downloaded https://github.com/apache/parquet-format/commit/40699d05bd24181de6b1457babbee2c16dce3803 and ran thrift.  Then I discovered the `regen.sh` and performed the steps therein. Yes, the `REVISION` in `regen.sh` should be updated. I'll make that part of a new PR for the format changes.

etseidl commented on parquet/src/file/metadata/mod.rs:
```diff
@@ -538,6 +553,9 @@ pub struct ColumnChunkMetaData {
     offset_index_length: Option<i32>,
     column_index_offset: Option<i64>,
     column_index_length: Option<i32>,
+    unencoded_byte_array_data_bytes: Option<i64>,
```
Sure. I should have thought to do that when I noticed the heap size tests failing after a merge.

etseidl replied:
done

etseidl commented on parquet/src/file/page_index/offset_index.rs:
```diff
@@ -0,0 +1,50 @@
+/// [`OffsetIndex`] information for a column chunk. Contains offsets and sizes for each page
+/// in the chunk. Optionally stores fully decoded page sizes for BYTE_ARRAY columns.
+#[derive(Debug, Clone, PartialEq)]
+pub struct ParquetOffsetIndex {
```
I wasn't sure about removing the typedef since that would be a breaking change, no? (This structure isn't yet user visible IIUC). I can feel September coming on fast, so if you're ok with breaking changes at this point, I could take a stab at revamping all of the page index read support as one of the break-out PRs.

etseidl commented:
@alamb I think I've addressed most of your concerns. Please take a look when you have a spare moment. I'm going to set this PR back to draft and begin on splitting into at least 2 more PRs (I've already submitted #6045 for the thrift component).

alamb commented:
> Certainly. I can submit the format changes first, but I'll keep this open for the time being while I address your comments. Should I open new issues for each PR, or is it sufficient to reference #5022 for all three?

I think using one ticket is just fine. No need to make new tickets

alamb commented:
Update: what do you think about doing incremental PRs to a feature branch? https://github.com/apache/arrow-rs/issues/6050

etseidl commented:
> Update: what do you think about doing incremental PRs to a feature branch? #6050

Sounds good to me. We can all use #6050 to discuss deconfliction in the event of conflicts. And we can afford to be a bit more free wheeling too 😄. For instance, I have a branch for changing the `offset_index` in `ParquetMetaData` to a `Vec<Vec<ParquetOffsetIndex>>`. I almost submitted that PR, but held back because the ripple of changes throughout the parquet code was a bit more than I expected. With the feature branch, I'll feel more comfortable going at a slower rate (add the variable length stats gathering, then change the API, then add histograms, then (perhaps) change the column index API to be consistent with the offset index. 

Thank you for helping to coordinate all this ❤️

etseidl commented:
Superseded by #6105 and others

alamb commented:
🎉
